### PR TITLE
[mongodb] supports pre- & post-image feature in mongo 6.0 (#2062)

### PR DESCRIPTION
**Pull request by yuxiqian**
This PR comes with the following changes:

* Updates `org.mongodb.kafka` and `mongodb-driver-sync` to fit MongoDB 6, and did minor fixes to handle breaking changes of dependencies.

* Adds a MongoDB CDC config `full.document.pre.post.image` to switch pre- and post-image retrieving feature on / off. Requires `scan.incremental.snapshot.enabled = true`.

* Generates normal `UPDATE BEFORE` and `DELETE` events when `full.document.pre.post.image = true`.

* Provides `UPDATE AFTER` events by post-image atomically when `full.document.pre.post.image = true`.

* Creates `MongoDBFullDocumentBeforeChangeITCase` to test its functionality.

> Declared a new dependency (`org.testcontainers:mongodb`) to implement `MongoDBContainer` that works with MongoDB 6. Original version was moved to `LegacyMongoDBContainer` and still used by `Legacy-` testcases.

### CODE REVIEW

**Comment by yuxiqian** [comment]
Done, marked ready for review.

Could @leonardBang help checking this? Thanks!

**Comment by leonardBang** [comment]
@Jiabao-Sun Would you like to review this PR when your time slot are available ?

**Comment by Jiabao-Sun** [comment]
> @Jiabao-Sun Would you like to review this PR when your time slot are available ?

OK

**Review by Jiabao-Sun — COMMENTED** [review]
Thanks @yuxiqian for this great works.
Most of it looks good to me, few comments were left.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/MongoDBSourceBuilder.java [hunk 1]
```diff
@@ -195,6 +195,17 @@ public MongoDBSourceBuilder<T> closeIdleReaders(boolean closeIdleReaders) {
         return this;
     }
 
+    /**
+     * full.document.pre.post.image
+     *
+     * <p>Whether to generate update before row data based on MongoDB 6.0 Full Document PrePostImage
+     * feature. <code>'scan.incremental.close-idle-reader.enabled'</code> needs to be set to true.
```
[the remark this answers is not in the document]
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/MongoDBSourceBuilder.java]: Added by mistake, removed.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceConfigFactory.java [hunk 2]
```diff
@@ -219,6 +220,18 @@ public MongoDBSourceConfigFactory closeIdleReaders(boolean closeIdleReaders) {
         return this;
     }
 
+    /**
+     * full.document.pre.post.image
+     *
+     * <p>Whether to generate update before row data based on MongoDB 6.0 Full Document Pre Post
+     * Image feature. <code>'scan.incremental.close-idle-reader.enabled'</code> needs to be set to
+     * true.
```
[the remark this answers is not in the document]
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceConfigFactory.java]: Removed.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableSource.java [hunk 3]
```diff
@@ -134,10 +136,21 @@ public MongoDBTableSource(
...
+        if (this.enableFullDocPrePostImage) {
+            // with FullDocPrePostImage feature
+            // U- row data can be emitted
+            return ChangelogMode.newBuilder()
+                    .addContainedKind(RowKind.INSERT)
+                    .addContainedKind(RowKind.UPDATE_BEFORE)
+                    .addContainedKind(RowKind.UPDATE_AFTER)
+                    .addContainedKind(RowKind.DELETE)
+                    .build();
+        }
```
[the remark this answers is not in the document]
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableSource.java:149]: Replaced with `ChangelogMode.all()` and `.upsert()`.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableSource.java [hunk 4]
```diff
@@ -198,7 +211,7 @@ public ScanRuntimeProvider getScanRuntimeProvider(ScanContext scanContext) {
                     .ifPresent(builder::heartbeatIntervalMillis);
             Optional.ofNullable(splitMetaGroupSize).ifPresent(builder::splitMetaGroupSize);
             Optional.ofNullable(splitSizeMB).ifPresent(builder::splitSizeMB);
-
+            Optional.of(enableFullDocPrePostImage).ifPresent(builder::enableFullDocPrePostImage);
```
[the remark this answers is not in the document]
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableSource.java]: Fixed.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableSource.java [hunk 5]
```diff
@@ -220,6 +233,11 @@ public ScanRuntimeProvider getScanRuntimeProvider(ScanContext scanContext) {
                                     + " is not supported by legacy source. To use this feature, 'scan.incremental.snapshot.enabled' needs to be set to true.");
             }
 
+            if (enableFullDocPrePostImage) {
+                throw new ValidationException(
+                        "Full Document pre and post image is not supported by legacy source. To use this feature, 'scan.incremental.snapshot.enabled' needs to be set to true.");
+            }
+
```
[the remark this answers is not in the document]
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableSource.java]: Fixed.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/MongoDBSource.java [hunk 6]
```diff
@@ -317,24 +337,24 @@ public DebeziumSourceFunction<T> build() {
 
             if (copyExisting != null) {
                 props.setProperty(
-                        MongoSourceConfig.COPY_EXISTING_CONFIG, String.valueOf(copyExisting));
+                        MongoSourceConfig.STARTUP_MODE_CONFIG,
+                        copyExisting ? "copy_existing" : "latest");
```
[the remark this answers is not in the document]
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/MongoDBSource.java:341]: Opened #2253 to address this.

**Comment by yuxiqian** [comment]
Comments addressed, cc @leonardBang

**Review by Jiabao-Sun — APPROVED** [review]
Thanks @yuxiqian for this great work.
LGTM.

@leonardBang do you have any suggetsions?

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/utils/MongoUtils.java [hunk 7]
```diff
@@ -339,8 +349,24 @@ public static <T> MongoCollection<T> collectionFor(
...
+                .toString();
+    }
+
     public static MongoClient clientFor(MongoDBSourceConfig sourceConfig) {
-        return MongoClientPool.getInstance().getOrCreateMongoClient(sourceConfig);
+        MongoClient client = MongoClientPool.getInstance().getOrCreateMongoClient(sourceConfig);
+        if (sourceConfig.isFullDocPrePostImageEnabled()) {
+            String mongoVersion = getMongoVersion(sourceConfig);
+            if (mongoVersion.charAt(0) < '6') {
+                throw new RuntimeException(
```
Comment (leonardBang) [inline flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/utils/MongoUtils.java:365]: Could we remove this validation logic to compile phase of the source? thus we can throw the exception earlier, and maybe we can even enable `full.document.pre.post.image` by default for mongo 6+ version after validation.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorDeserializationSchema.java [hunk 8]
```diff
@@ -129,29 +129,56 @@ public void deserialize(SourceRecord record, Collector<RowData> out) throws Exce
                 extractBsonDocument(value, valueSchema, MongoDBEnvelope.DOCUMENT_KEY_FIELD);
         BsonDocument fullDocument =
                 extractBsonDocument(value, valueSchema, MongoDBEnvelope.FULL_DOCUMENT_FIELD);
-
+        BsonDocument fullDocumentBeforeChange =
+                extractBsonDocument(
+                        value, valueSchema, MongoDBEnvelope.FULL_DOCUMENT_BEFORE_CHANGE_FIELD);
```
Comment (leonardBang) [inline flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorDeserializationSchema.java:134]: IIUC, we need to extract this document when `full.document.pre.post.image` config is enabled. We can avoid the unnecessary cost for lower mongo version if we know the config.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceOptions.java [hunk 9]
```diff
@@ -133,4 +133,12 @@ public class MongoDBSourceOptions {
                     .defaultValue(64)
                     .withDescription(
                             "The chunk size mb of incremental snapshot. Defaults to 64mb.");
+
+    @Experimental
+    public static final ConfigOption<Boolean> FULL_DOCUMENT_PRE_POST_IMAGE =
+            ConfigOptions.key("full.document.pre.post.image")
```
Comment (leonardBang) [inline flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceOptions.java:139]: `document`, `pre-image` and `post-image` are MongoDB concepts, when defined this option in Flink Connector/SQL, I prefer to use changelog related concept for Flink users, how about `scan.full.changelog` or something else?

**Review by leonardBang — COMMENTED** [review]
Thanks @yuxiqian for the nice work, I left some comments.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/internal/MongoDBConnectorSourceTask.java [hunk 10]
```diff
@@ -71,17 +70,28 @@ public class MongoDBConnectorSourceTask extends SourceTask {
 
     public static final String COLLECTION_INCLUDE_LIST = "collection.include.list";
 
+    public static final String STARTUP_MODE_COPY_EXISTING_MAX_THREADS_CONFIG =
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/internal/MongoDBConnectorSourceTask.java:73]: +1 to remove all `copy.existing` with STARTUP MODE

[the remark this answers is not in the document]
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/internal/MongoDBConnectorSourceTask.java]: Renamed Table API config `copy.existing.queue.size` to `initial.snapshotting.queue.size`.

Created DataStream API SourceBuilder config `initialSnapshottingMaxThreads`, `initialSnapshottingQueueSize`, and `initialSnapshottingPipeline`. Also deprecated `copyExisting-` ones.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/utils/MongoUtils.java (hunk 7 above)
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/utils/MongoUtils.java]: Removed.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorDeserializationSchema.java (hunk 8 above)
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorDeserializationSchema.java]: Fixed by passing `scan.full.changelog` config to the deserializer.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceOptions.java (hunk 9 above)
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceOptions.java]: Renamed option to `scan.full.changelog` / `.scanFullChangelog`.

Comment (leonardBang) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceOptions.java]: `scan.full-changelog` may be better when compared to other flink connector options

Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceOptions.java]: Done.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorDeserializationSchema.java [hunk 11]
```diff
@@ -110,12 +113,14 @@ public MongoDBConnectorDeserializationSchema(
...
             TypeInformation<RowData> resultTypeInfo,
-            ZoneId localTimeZone) {
+            ZoneId localTimeZone,
+            boolean preImageEnabled) {
         this.hasMetadata = checkNotNull(metadataConverters).length > 0;
         this.appendMetadataCollector = new AppendMetadataCollector(metadataConverters);
         this.physicalConverter = createConverter(physicalDataType);
         this.resultTypeInfo = resultTypeInfo;
         this.localTimeZone = localTimeZone;
+        this.hasPreImageData = preImageEnabled;
```
Comment (leonardBang) [inline flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorDeserializationSchema.java:123]: we can introduce builder pattern for compatibility consideration here

Comment (leonardBang) [inline flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorDeserializationSchema.java:123]: I mean we can build different runtime converter  like `DeserializationRuntimeConverter`  (you can name a new one) for `hasPreImageData` enabled  case or  `hasPreImageData` disabled case during compile phase(i.e. constructor), and then we do not need to judge `hasPreImageData` per-record in deserialize method.

**Review by leonardBang — COMMENTED** [review]
tips: when address comments or sync the latest master branch, `git rebase master` is recommended than `git merge master` as it's more friendly to reviewer and maintainer.

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorDeserializationSchema.java (hunk 11 above)
Comment (yuxiqian) [reply flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorDeserializationSchema.java]: Created `MongoDBConnectorFullChangelogDeserializationSchema` class and construct corresponding deserializer in `getScanRuntimeProvider`.

**Comment by yuxiqian** [comment]
Since this PR is getting too big already, I've created another PR #2267 to address duplicate snapshot records issue.

cc @leonardBang

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorFullChangelogDeserializationSchema.java [hunk 12]
```diff
@@ -0,0 +1,820 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
```
Comment (leonardBang) [inline flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorFullChangelogDeserializationSchema.java:2]: minor: incorrect header

File: flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorFullChangelogDeserializationSchema.java [hunk 13]
```diff
@@ -0,0 +1,820 @@
...
+            ZoneId localTimeZone) {
+        this.hasMetadata = checkNotNull(metadataConverters).length > 0;
+        this.appendMetadataCollector = new AppendMetadataCollector(metadataConverters);
+        this.physicalConverter = createConverter(physicalDataType);
+        this.resultTypeInfo = resultTypeInfo;
+        this.localTimeZone = localTimeZone;
+    }
+
+    @Override
+    public void deserialize(SourceRecord record, Collector<RowData> out) throws Exception {
```
Comment (leonardBang) [inline flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorFullChangelogDeserializationSchema.java:122]: We can let `MongoDBConnectorFullChangelogDeserializationSchema` extends `MongoDBConnectorDeserializationSchema`  and then  overrides this method to reuse common code pice.

**Review by leonardBang — COMMENTED** [review]
Thanks @yuxiqian for the update, the PR looks good, I only left two minor comments.

**Comment by yuxiqian** [comment]
Thanks for @leonardBang's review, comments addressed.

**Review by leonardBang — APPROVED** [review]
+1

**Pull request merged**